Help Menu Functionality for Components and Paradigms

lefevre-fraser opened the pull request:
Registry keys will be used to provide help documentation for paradigms and components.
Registry string values may be a website, local HTML file, or basically any file type.
If the registry key is set then a help menu will be visible on the main menu bar.

For Paradigms: (Main Menu)>Help>ParadigmName>(Online/Offline Help)
For Components: (Main Menu)>Help>Component Help>ComponentName>(Online/Offline Help)

Component Registry Keys: 
(HKEY_CURRENT_USER/HKEY_LOCAL_MACHINE)>SOFTWARE>WOW6432NODE>GME>Components>ComponentName>OnlineHelp
(HKEY_CURRENT_USER/HKEY_LOCAL_MACHINE)>SOFTWARE>WOW6432NODE>GME>Components>ComponentName>OfflineHelp

Paradigm Registry Keys:
(HKEY_CURRENT_USER/HKEY_LOCAL_MACHINE)>SOFTWARE>WOW6432NODE>GME>Paradigms>ParadigmName>{ParadigmVersionUUID}>OnlineHelp
(HKEY_CURRENT_USER/HKEY_LOCAL_MACHINE)>SOFTWARE>WOW6432NODE>GME>Paradigms>ParadigmName>{ParadigmVersionUUID}>OfflineHelp

lefevre-fraser committed:
Help Menu Functionality for Components and Paradigms

ksmyth commented on GME/Gme/GME.rc:
```diff
@@ -1131,14 +1156,15 @@ BEGIN
 VS_VERSION_INFO VERSIONINFO
- FILEVERSION GME_VERSION_RC
- PRODUCTVERSION GME_VERSION_RC
+ FILEVERSION 11,11,3,0
```
Please revert this unintentional changes

lefevre-fraser replied:
Done, did you notice any others? I found one, and updated it as well.

ksmyth commented on GME/Gme/GMEApp.cpp:
```diff
@@ -124,8 +125,12 @@ BEGIN_MESSAGE_MAP(CGMEApp, CWinAppEx)
 	ON_COMMAND_RANGE(ID_FILE_INTERPRET1, ID_FILE_INTERPRET_LAST, OnRunInterpreter)
-	ON_UPDATE_COMMAND_UI_RANGE( ID_FILE_RUNPLUGIN1, ID_FILE_RUNPLUGIN_LAST, OnUpdateFilePluginX)
-	ON_UPDATE_COMMAND_UI_RANGE( ID_FILE_INTERPRET1, ID_FILE_INTERPRET_LAST, OnUpdateFileInterpretX)
+	ON_COMMAND(ID_HELP_COMPONENTHELP, DummyCommand)
```
Was this change intentional?

lefevre-fraser replied:
Yes, the "dummy command" is needed so that the popup menu can be enabled.
If it can be made to work without it that would be fine. I could not get it working without it.

ksmyth commented on GME/Gme/GMEApp.cpp:
```diff
@@ -2972,7 +3170,73 @@ void CGMEApp::OnRunInterpreter(UINT nID) {
+		COMTHROW(mgaProject->get_MetaVersion(&pver));
+		COMTHROW(mgaProject->CommitTransaction());
+	}
+	std::wstringstream wss;
```
This is weird. Wouldn't `static_cast<const wchar_t*>(pname)` work here?

lefevre-fraser replied:
I just tried that. visual studio complained: "expression must have integral or unscoped enum type"
I found this works though:
CGMEEventLogger::LogGMEEvent((_T("CGMEApp::OnRunParadigmHelp ")
		+ std::wstring(static_cast<const wchar_t*>(pname))
		+ _T("\r\n")).c_str());

lefevre-fraser replied:
I just commited&pushed the change

ksmyth commented on GME/Gme/GMEApp.cpp:
```diff
@@ -3113,6 +3377,15 @@ void CGMEApp::OnUpdateFileInterpretX(CCmdUI* pCmdUI)
+	//	|| m_vecDisabledComps.find(pCmdUI->m_nID) == m_vecDisabledComps.end();
+	//MessageBox(NULL, (L"Enabling Help Command" + std::to_wstring(pCmdUI->m_nID)).c_str(), L"", MB_OK);
+	if (pCmdUI->m_nID >= ID_HELP_COMPONENTHELP_FIRST && pCmdUI->m_nID <= ID_HELP_COMPONENTHELP_LAST)
+		pCmdUI->Enable(true/*enabled*/);
```
Should be `TRUE`

lefevre-fraser replied:
Ok, I took care of it

lefevre-fraser commented on GME/Gme/GME.rc:
```diff
@@ -395,7 +393,9 @@ BEGIN
-    BEGIN
+	BEGIN
+		MENUITEM "&Paradigm Help",              ID_HELP_PARADIGMHELP
+		MENUITEM "C&omponent Help",             ID_HELP_COMPONENTHELP
```
There is some "phantom"? white-space

ksmyth commented on GME/Gme/GMEView.cpp:
```diff
@@ -468,6 +468,7 @@ BEGIN_MESSAGE_MAP(CGMEView, CScrollZoomView)
 	ON_UPDATE_COMMAND_UI(ID_FILE_CHECK, OnUpdateFileCheck)
 	ON_UPDATE_COMMAND_UI(ID_FILE_CHECKSEL, OnUpdateFileCheckSelected)
 	ON_UPDATE_COMMAND_UI(ID_FILE_SETTINGS, OnUpdateFileSettings)
+	ON_UPDATE_COMMAND_UI(ID_HELP_CONTENTS, OnUpdateFileSettings)
```
Is this necessary?

lefevre-fraser replied:
Yes, That is what triggers the update to the help menu lists

ksmyth replied:
If I comment it out, it still works. Is only the entry in MainForm needed?

lefevre-fraser replied:
You're right. 
I thought it was needed because the toolbar is different when a part/component/testbench is open, which I thought was only controlled by GMEView.

ksmyth commented on GME/Gme/GMEApp.cpp:
```diff
@@ -2972,7 +3115,62 @@ void CGMEApp::OnRunInterpreter(UINT nID) {
+		_T("\r\n"));
+
+	// Focus must be killed to flush ObjectInspector and Browser
+	::SetFocus(NULL);
```
Is this necessary?

lefevre-fraser replied:
I don't know. It was part of the code I coppied from the OnRunComponent function.
Are you just refering to the `::SetFocus(NULL);` line?

lefevre-fraser replied:
I got rid of it. There was no change in functionality.

ksmyth committed:
Help Menu Pull Request 1 - Kevin Smyth

Component help: use tooltip names
Help menu: fewer transactions
Paradigm help: fix HKCU paradigms
